tom/dependencies.py

```diff
@@ -235,6 +266,93 @@ def update_single_dep(self, dep):
+        upstream_name = 'cfengine'
+        local_path = "../" + repo_name
+        self.buildscripts = GitRepo(local_path, repo_name, upstream_name, self.username, 'master')
+
+        branches = branches.split(',')
+
+        # fetch versions from all branches
+        deps_table = {}
+        # deps_table is a 2d dict: deps_table[dep][branch]=version
+        branch_column_widths = {}
+        for branch in branches:
+            branch_column_widths[branch] = len(branch)
+            self.buildscripts.checkout(branch)
+            self.buildscripts.run_command('pull')
+            deps_versions = self.collect_deps(branch)
+            # deps_versions is a dict: deps_versions[dep]=version
+            for dep in deps_versions:
+                if not dep in deps_table:
+                    deps_table[dep] = {}
+                deps_table[dep][branch] = deps_versions[dep]
+                branch_column_widths[branch] = max(branch_column_widths[branch], len(deps_versions[dep]))
+
+        # patch the readme
+        self.buildscripts.checkout('master')
+        self.readme_file_path = 'deps-packaging/README.md'
+        readme_file = self.buildscripts.get_file(self.readme_file_path)
+        readme_lines = readme_file.split('\n')
+        has_notes = False # flag to say that we're in a table that has "Notes" column
+        for i, line in enumerate(readme_lines):
+            if not line.startswith('| '):
+                continue
+            if line.startswith('| CFEngine version '):
+                has_notes = 'Notes' in line
+                branches_text = ' | '.join([branch.ljust(branch_column_widths[branch]) for branch in branches])
+                line = '| CFEngine version | ' + branches_text + (' | Notes |' if has_notes else ' |')
+            elif line.startswith('|  --'):
+                branches_text = ' | '.join(['-'*branch_column_widths[branch] for branch in branches])
+                line = '|  --------------  | ' + branches_text + (' | ----- |' if has_notes else ' |')
+            else:
+                dep = re.match('\\|  ([a-z0-9-]*) ', line, flags=re.IGNORECASE)
+                if dep:
+                    dep = dep.group(1)
+                else:
+                    log.warn("didn't find dep in line [%s]", line)
+                    continue
+                if dep not in deps_table:
+                    log.warn("unknown dependency in README: [%s] line [%s], will be EMPTY", dep, line)
+                    deps_table[dep] = {}
+                if has_notes:
+                    note = re.search('\| [^|]* \|$', line)
+                    if not note:
+                        log.warn("didn't find note in line [%s]", line)
+                        note = '|  |'
+                    note = note.group(0) # group(0) is full matched string
+                branches_text = ' | '.join([(deps_table[dep][branch] if branch in deps_table[dep] else '-').ljust(branch_column_widths[branch]) for branch in branches])
```

Comment:
You can drop the `[ ]` inside the `join()` call. And I would wrap it before the `for`. And maybe use [`defaultdict`](https://docs.python.org/3.7/library/collections.html#collections.defaultdict) for `deps_table` like this:

```
deps_table = defaultdict(lambda: "-")
```
and then you can drop the `if-else`.